thesis/03-implementation.tex

```diff
@@ -83,23 +83,491 @@ \subsubsection{Backend}
+
+
+\subsubsection{Back-end}
+
+\begin{figure}[H]
+  \caption{The Community application consists of two main sub-applications.}
+  \label{3-backend-community-components}
+  \centering
+    \includegraphics[width=\textwidth]{assets/3-community-application.png}
+\end{figure}
+
+The community component is built using Django framework and split into a couple of sub-applications:
+
+\begin{description}
+	\item[accounts] - responsible for functionality related to user management. 
+	\item[guides] - responsible for controlling access and running all business logic related to Instruction and Animation files.
+\end{description}
+
+\textbf{Guides}
+
+Guides is the main sub-application. It is responsible for handling guide creation, deletion, marking a guide as a favorite, and delegation of Guide processing tasks.
+
+The sub-application contains two sub-components:
+
+\begin{itemize}
+	\item a processing task definition
+	\item a web application
+\end{itemize}
+
+The only entity defined in the Guides component is a Guide itself.
+
+\begin{figure}[H]
+  \caption{Guides - ERD diagram}
+  \centering
+    \includegraphics[width=0.6\textwidth]{assets/3-erd-guides.png}
+\end{figure}
+
+The two most important fields include: \begin{description}
+	\item[guide\_file] - an Instruction, a set of Steps.
+	\item[animation\_file] - an Animated Guide, a set of Transitions.
+\end{description}
+
+When a Guide is created, it is scheduled onto a task queue and later assembled into an Animated Guide. The processing task invokes Solver to calculate intermediate steps forming a smooth folding animation. 
+
+\medskip 
+Guide processing may terminate with 3 different outcomes.
+
+\begin{itemize}
+	\item An Instruction may be impossible to fold programmatically or realistically. It might occur that the solving algorithm never stops, consuming all available resources. To counteract, a hard timeout has been introduced after which the Guide is marked as timed out - \texttt{TMO}.
+	\item An Instruction file may be malformed or unprocessable, in such scenario the Guide is marked as errored - \texttt{ERR}.
+	\item Finally, a properly folded Instruction is marked as done - \texttt{DON}, and the animation file is persisted in the file system.
+\end{itemize}
+
+
+\begin{figure}[H]
+  \caption{Guide states}
+  \centering
+    \includegraphics[width=0.8\textwidth]{assets/3-guide-states.png}
+\end{figure}
+
+
+
+\medskip
+
+\textbf{Accounts}
+
+Accounts sub-application's responsibilities include, but are not limited to:
+\begin{itemize}
+	\item authentication
+	\item account creation
+	\item password resetting 
+	\item password changing
+\end{itemize}
+
+To fulfill its duties, the Accounts sub-application exposes several REST API endpoints.
+The data model is mostly derived from the built-in Django model increasing the development speed considerably.
```

Comment:
increasing -> which increased

It increased the speed once, it does not increase it continuously